### ISSUE TITLE
documentation of Adafactor does not match the implementation

### ISSUE THREAD

**Issue by yorkerlin** [pytorch/pytorch#154862]
### 📚 The doc issue

Hi, I have one question about the implementation of Adafactor. 

According to the PyTorch document and the [Adafactor paper](https://arxiv.org/pdf/1804.04235), a `sum` of the gradient square should be used. 
However, the existing implementations use a `mean` of the gradient square. 

The `mean` of the gradient square is used in the implementations.
https://github.com/pytorch/pytorch/blob/v2.7.0/torch/optim/_adafactor.py#L538
https://github.com/pytorch/pytorch/blob/v2.7.0/torch/optim/_adafactor.py#L547
https://github.com/facebookresearch/fairseq/blob/main/fairseq/optim/adafactor.py#L236
https://github.com/google-research/t5x/blob/main/t5x/adafactor.py#L524

The `sum` of the gradient square is used in the PyTorch document and Adafactor paper.
https://docs.pytorch.org/docs/stable/generated/torch.optim.Adafactor.html
https://arxiv.org/pdf/1804.04235

Why is this discrepancy not documented or discussed?

Reference:
https://github.com/pytorch/pytorch/issues/109581

@janeyx99 

cc @svekars @sekyondaMeta @AlannaBurke @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by janeyx99** [comment]
Haha I remember having the same question when I was writing this implementation before too! I remember working out the math and realizing it's the same thing (when you look at the whole implementation)--a mean is just a scaled sum and is usually smaller, which is advantageous to be able to represent in the limited numerical ranges.

**Comment by yorkerlin** [comment]
My understanding is that the same `beta2` should be used for all blocks (e.g. 2d tensors)
The `mean` version does not guarantee that the same `beta2` is used for all blocks (e.g., 2d tensors).

**Comment by yorkerlin** [comment]
@janeyx99 

For block 1
C_1 = (1-\beta2) C_1 + \beta2 `sum`(G^2, dim=1) <=> C_1 / dim(R_1)  =  (1-\beta2)  C_1 / dim(R_1) + \beta2 `mean`(G^2, dim=1) 
R_1 = (1-\beta2) C_1 + \beta2 `sum`(G^2, dim=0) <=>  R_1 / dim(C_1)=  (1-\beta2) R_1 / dim(C_1)  + \beta2 `mean`(G^2, dim=0) 

For block 2
C_2 = (1-\beta2) C_2 + \beta2 sum(G^2 dim=1) <=>  C_2 / dim(R_2) =  (1-\beta2) C_2 / dim(R_2) + \beta2 `mean`(G^2, dim=1) 
R_2 = (1-\beta2) C_2 + \beta2 sum(G^2, dim=0) <=>  R_2 / dim(C_2)   =  (1-\beta2)  R_2 / dim(C_2)  + \beta2 `mean`(G^2, dim=0) 


So, the `mean` version implicitly introduces a block-wise learning rate for preconditioning since $dim(R_1) dim(C_1) \neq dim(R_2) dim(C_2)$ in general.

**Comment by janeyx99** [comment]
The scale factor gets removed with the division with device_row_vars here https://github.com/pytorch/pytorch/blob/v2.7.0/torch/optim/_adafactor.py#L561

**Comment by yorkerlin** [comment]
Thanks for the explanation!  I think it's better to document this, as it may not be evident to others.

**Comment by janeyx99** [comment]
I'd review a PR adding a description for this

**Comment by morrison-turnansky** [comment]
I'll work on this.
